### ISSUE TITLE
Option to extract all of the type-only imports as a separate group

### ISSUE THREAD

**Issue by ekowalsk** [lydell/eslint-plugin-simple-import-sort#61]
Hi!

Thank you for your nice plugin and all of the effort you’ve put in it! 

What makes it distinctive from other plugins for sorting imports is handling the type-only imports (`import type` syntax).
It would be really amazing, if your plugin enabled extracting all of that type-only imports into a separate group - in case of big projects it would really facilitate organizing imports. Or at least - if it would have an option to put all of the type-only imports at the beginning of specified groups.
I hope you will consider my suggestion :)

Thank you!

**Comment by lydell** [comment]
Great idea! Just like it’s possible to customize where side effect imports go we should make it possible to customize where type imports go.

**Comment by ekowalsk** [comment]
> Great idea! Just like it’s possible to customize where side effect imports go we should make it possible to customize where type imports go.

Thank you! I'm very happy to hear that!

**Comment by lydell** [comment]
Could you post the imports from 1-3 files you have in your desired order? It helps when designing the feature and when writing test cases.

**Comment by ekowalsk** [comment]
Here goes links for example files:
[](https://github.com/polkadot-js/apps/blob/master/packages/react-components/src/Event.tsx)
[](https://github.com/polkadot-js/apps/blob/master/packages/apps/src/Endpoints/Group.tsx)
[](https://github.com/polkadot-js/apps/blob/master/packages/react-components/src/CardSummary.tsx)

**Comment by lydell** [comment]
Awesome! Can you post your desired order as well?

**Comment by jacogr** [comment]
The above is actually via a slightly patched version of this plugin... :) So the ordering there is as intended. (Originally we had all the types in a separate grouping, but that overly-complicates everything). 

Basically the hack we applied (stopgap) for the above output is the following -

- https://github.com/polkadot-js/apps/blob/master/patches/eslint-plugin-simple-import-sort%2B6.0.1.patch (basically iy just moved the sort first on types)

The config is quite simple - https://github.com/polkadot-js/apps/blob/master/.eslintrc.js#L31-L39

**Comment by lydell** [comment]
I’ve published 7.0.0-beta.1 with support for choosing where type imports go. Type imports have `\u0000` at the _end_  (similar to how side effect import have `\u0000` at the beginning).

You should be able to use something like this:

```json
{
  groups: [
    ["^[^@.].*\\u0000$", "^[^/.]"],
    ["^@polkadot.*\\u0000$", "^@polkadot"],
    ["^\\..*\\u0000$", "^\\."]
  ]
}
```

Could you try it out?

**Comment by jacogr** [comment]
Perfect, thank you! We will give it a run through tomorrow.

**Comment by ekowalsk** [comment]
Thank you so much! 😀  Definitely we will give feedback tomorrow!

**Comment by ekowalsk** [comment]
> I’ve published 7.0.0-beta.1 with support for choosing where type imports go. Type imports have `\u0000` at the _end_ (similar to how side effect import have `\u0000` at the beginning).
> 
> You should be able to use something like this:
> 
> ```json
> {
>   "groups": [
>     ["^[^@.].*\\u0000$", "^[^/.]"],
>     ["^@polkadot.*\\u0000$", "^@polkadot"],
>     ["^\\..*\\u0000$", "^\\."]
>   ]
> }
> ```
> 
> Could you try it out?

The extracting type-only imports feature into separate groups works perfect!! I've applied that configuration: 
groups: [ 
            ['^.*\u0000$'],
            ['^[^\/\\.]'],
            ['^@polkadot'],
            ['^\\.']
        ],
The only thing I noticed is that it didn't handle properly the comment for the `punycode` import in that file https://github.com/polkadot-js/apps/blob/master/packages/apps/src/Endpoints/index.tsx , while the previous version did.

**Comment by lydell** [comment]
Glad to hear it’s working, and that you could work out some config on your own!

> The only thing I noticed is that it didn't handle properly the comment for the punycode import in that file https://github.com/polkadot-js/apps/blob/master/packages/apps/src/Endpoints/index.tsx , while the previous version did

I would be surprised if anything related to comments has changed in 7.0.0-beta.1, because the whole “`\u0000` at the end for type imports” was a super small change in a different file than the comments handling.

I think this happened – with your new config, type imports are in a group of their own _before_ all other imports (except side effect imports). So when autofixing the top of the imports chunk changed from:

```js
/* eslint-disable-next-line node/no-deprecated-api */
import punycode from 'punycode';
```

To:

```js
/* eslint-disable-next-line node/no-deprecated-api */
import type { Whatever } from 'something';
```

So the `eslint-disable-next-line` became non-effective and you had to manually move it back to the `punycode` import.

It’s unfortunately not possible to know if a comment before the first import belongs to the import or to the whole file, so that’s why it isn’t moved. See https://github.com/lydell/eslint-plugin-simple-import-sort#comment-and-whitespace-handling for more information.

**Comment by jacogr** [comment]
This feature rocks! Thanks a lot.

**Comment by lydell** [comment]
Released in v7.0.0. Thanks for all the help!
